regression: cycle in MIR opts

Found by 1.56 crater -- https://crater-reports.s3.amazonaws.com/beta-1.56-2/beta-2021-09-12/gh/jkarneges.rust-async-bench/log.txt

```
[INFO] [stdout] error[E0391]: cycle detected when optimizing MIR for `run::listen::{closure#0}`
[INFO] [stdout]    --> src/run.rs:167:28
[INFO] [stdout]     |
[INFO] [stdout] 167 |   ) -> Result<(), io::Error> {
[INFO] [stdout]     |  ____________________________^
[INFO] [stdout] 168 | |     let listener = AsyncFakeListener::new(reactor, stats);
[INFO] [stdout] 169 | |
[INFO] [stdout] 170 | |     for _ in 0..CONNS_MAX {
[INFO] [stdout] ...   |
[INFO] [stdout] 180 | |     Ok(())
[INFO] [stdout] 181 | | }
[INFO] [stdout]     | |_^
[INFO] [stdout]     |
[INFO] [stdout]     = note: ...which requires computing layout of `impl std::future::Future`...
[INFO] [stdout]     = note: ...which requires computing layout of `std::future::from_generator::GenFuture<[static generator@src/run.rs:213:3: 218:2]>`...
[INFO] [stdout]     = note: ...which requires computing layout of `[static generator@src/run.rs:213:3: 218:2]`...
[INFO] [stdout] note: ...which requires optimizing MIR for `run::do_async::{closure#0}`...
[INFO] [stdout]    --> src/run.rs:213:3
[INFO] [stdout]     |
[INFO] [stdout] 213 |   ) {
[INFO] [stdout]     |  ___^
[INFO] [stdout] 214 | |     match invoke {
[INFO] [stdout] 215 | |         AsyncInvoke::Listen => listen(spawn, ctx, reactor, stats).await.unwrap(),
[INFO] [stdout] 216 | |         AsyncInvoke::Connection(stream) => connection(stream).await.unwrap(),
[INFO] [stdout] 217 | |     }
[INFO] [stdout] 218 | | }
[INFO] [stdout]     | |_^
[INFO] [stdout]     = note: ...which requires computing layout of `impl std::future::Future`...
[INFO] [stdout]     = note: ...which requires computing layout of `std::future::from_generator::GenFuture<[static generator@src/run.rs:167:28: 181:2]>`...
[INFO] [stdout]     = note: ...which requires computing layout of `[static generator@src/run.rs:167:28: 181:2]`...
[INFO] [stdout]     = note: ...which again requires optimizing MIR for `run::listen::{closure#0}`, completing the cycle
```

cc @oli-obk

--- a/compiler/rustc_mir_transform/src/remove_zsts.rs
+++ b/compiler/rustc_mir_transform/src/remove_zsts.rs
@@ -9,6 +9,10 @@ pub struct RemoveZsts;
 
 impl<'tcx> MirPass<'tcx> for RemoveZsts {
     fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        // Avoid query cycles (generators require optimized MIR for layout).
+        if tcx.type_of(body.source.def_id()).is_generator() {
+            return;
+        }
         let param_env = tcx.param_env(body.source.def_id());
         let (basic_blocks, local_decls) = body.basic_blocks_and_local_decls_mut();
         for block in basic_blocks.iter_mut() {

--- a/src/test/ui/mir/remove-zsts-query-cycle.rs
+++ b/src/test/ui/mir/remove-zsts-query-cycle.rs
@@ -1,0 +1,16 @@
+// Regression test for #88972. Used to cause a query cycle:
+//   optimized mir -> remove zsts -> layout of a generator -> optimized mir.
+//
+// edition:2018
+// compile-flags: --crate-type=lib
+// build-pass
+
+pub async fn listen() -> Result<(), std::io::Error> {
+    let f = do_async();
+    std::mem::forget(f);
+    Ok(())
+}
+
+pub async fn do_async() {
+    listen().await.unwrap()
+}

